testsuite/classlibrary/TestCoreUGens.sc

```diff
@@ -309,124 +318,143 @@ TestCoreUGens : UnitTest {
+			server.sync;
 		};
 
-		condvar.waitFor(1, { completed == tests.size });
+		testsFinished = cond.waitFor((1.5 * tests.size * testDur).max(minTimeOut), { completed == tests.size });
+		this.assert(testsFinished, "TIMEOUT: ugen_generator_equivalences tests", report: false);
 	}
 
 	test_exact_convergence {
-		var n, v;
-		var condvar = CondVar();
+		var cond = CondVar();
 		var completed = 0;
+		var n, v, tests, testDur, testsFinished;
 
 		// Tests for things that should converge exactly to zero
-		var tests = Dictionary[
+		tests = Dictionary[
+
 			//////////////////////////////////////////
 			// Pan2 amplitude convergence to zero test, unearthed by JH on sc-dev 2009-10-19.
-			"Pan2.ar(ar, , kr) should converge properly to zero when amp set to zero" -> {(Line.ar(1,0,0.2)<=0)*Pan2.ar(BrownNoise.ar, 0, Line.kr(1,0, 0.1)>0).mean},
+			"Pan2.ar(ar, 0, kr) should converge properly to zero when amp set to zero" -> {
+				(Line.ar(1,0,0.2)<=0)*Pan2.ar(BrownNoise.ar, 0, Line.kr(1,0, 0.1)>0).mean
+			},
 
+			// add more tests here
 		];
+
 		server.bootSync;
-		tests.keysValuesDo{|name, func|
-			func.loadToFloatArray(1, server, { |data|
-				this.assertArrayFloatEquals(data, 0, name.quote, within: 0.0, report: true);
+		testDur = defaultNumBlocksToTest * server.options.blockSize / server.sampleRate;
+
+		tests.keysValuesDo{ |text, func|
+			func.loadToFloatArray(testDur, server, { |data|
+				this.assertArrayFloatEquals(data, 0, text.quote, within: 0.0, report: true);
 				completed = completed + 1;
-				condvar.signalOne;
+				cond.signalOne;
 			});
-			rrand(0.05, 0.1).wait;
+			server.sync;
 		};
-
-		condvar.waitFor(1, { completed == tests.size });
+		"timing out after % sec\n".postf(tests.size * testDur);
+		testsFinished = cond.waitFor((1.5 * tests.size * testDur).max(minTimeOut), { completed == tests.size });
+		this.assert(testsFinished, "TIMEOUT: exact_convergence tests", report: false);
 	}
 
+
 	test_muladd {
-		var n, v;
-		var condvar = CondVar();
+		var cond = CondVar();
 		var completed = 0;
+		var n, v, tests, testDur, testsFinished;
 
-		var tests = Dictionary[
-		];
+		tests = Dictionary[];
 		[[\ar,\kr], [2,0,5], [\ar,\kr], [2,0,5], [\ar,\kr], [2,0,5]].allTuples.do{|tup|
-			//tup.postln;
 			tests["%%.madd(%%, %%)".format(*tup)] =
 			"{DC.%(%).madd(DC.%(%), DC.%(%)) - (% * % + %)}".format(*(tup ++ tup[1,3..])).interpret;
 		};
 
 		server.bootSync;
+		testDur = defaultNumBlocksToTest * server.options.blockSize / server.sampleRate;
+
 		tests.keysValuesDo{|name, func|
-			func.loadToFloatArray(0.01, server, { |data|
+			func.loadToFloatArray(testDur, server, { |data|
 				this.assertArrayFloatEquals(data, 0, name.quote, report: true);
 				completed = completed + 1;
-				condvar.signalOne;
+				cond.signalOne;
 			});
+			server.sync;
 		};
 
-		condvar.waitFor(1, { completed == tests.size });
+		testsFinished = cond.waitFor((1.5 * tests.size * testDur).max(minTimeOut), { completed == tests.size });
```

Comment:
I just had a question about your intent here. The time here is the time between each `condvar.signalOne`, not the total time of waiting (documentation seems lacking here), is it correct to multiply by the number of tests in this case  — I mean it works so it doesn't really matter.

Reply:
Thanks for looking at this in detail. The way to think about this is the timeout is set to a duration beyond the completion time of _all_ tests. It's a bit confusing because this condition is signaled after each test, but it's actually waiting for the predicate `completed == tests.size` to be `true`. From the [docs](https://docs.supercollider.online/Classes/CondVar.html#-waitFor):

>If predicate is not `nil`, this method returns immediately if `predicate.value` is `true`, and otherwise blocks either until the timeout expires, or until the thread is woken by `signalOne` or `signalAll` and `predicate.value` is `true`.

The mechanics of `CondVar` take some getting used to (I have to re-read docs every time I use it).

Reply:
Thanks for this! I actually wrote some tests before I wrote that to try and make sure what I was saying was correct... I guess I messed up somewhere as I can't repeat it now! Thanks again!